lib/Conversion/RTLToLLHD/RTLToLLHD.cpp

```diff
@@ -207,39 +272,54 @@ struct ConvertInstance : public OpConversionPattern<InstanceOp> {
 
       Location loc = result.getLoc();
 
-      Value init;
+      // Since we need to have a signal for this result, see if an OutputOp maps
+      // it to an output signal of our parent module. In that case we can just
+      // use that signal.
+      Value sig;
       for (auto &use : result.getUses()) {
-        Value sig;
-
         if (isa<OutputOp>(use.getOwner())) {
-          // If the result is an entity output, use that signal.
           auto entity = instance->getParentOfType<EntityOp>();
+          if (!entity)
+            continue;
           sig = entity.getArgument(entity.ins() + use.getOperandNumber());
-        } else {
-          // Otherwise, materialize a signal.
-          if (!init)
-            init = rewriter.create<ConstOp>(
-                loc, resultType, rewriter.getIntegerAttr(resultType, 0));
-
-          SmallString<8> sigName(instance.instanceName());
-          sigName += "_result_";
-          sigName += std::to_string(result.getResultNumber());
-          sig = rewriter.createOrFold<SigOp>(loc, SigType::get(resultType),
-                                             sigName, init);
+          break;
         }
+      }
 
-        // Replace all uses of this result with the signal.
-        rewriter.updateRootInPlace(use.getOwner(), [&]() { use.set(sig); });
+      // Otherwise materialize a signal.
+      // TODO: This should be a `llhd.buffer` operation.
+      if (!sig) {
+        auto init = rewriter.create<ConstOp>(
+            loc, resultType, rewriter.getIntegerAttr(resultType, 0));
+        SmallString<8> sigName(instance.instanceName());
+        sigName += "_result_";
+        sigName += std::to_string(result.getResultNumber());
+        sig = rewriter.createOrFold<SigOp>(loc, SigType::get(resultType),
+                                           sigName, init);
+      }
 
-        results.push_back(sig);
+      // Make OutputOps directly refer to this signal, which allows them to use
+      // a ConnectOp rather than a PrbOp+DrvOp combo.
+      for (auto &use : llvm::make_early_inc_range(result.getUses())) {
+        if (isa<OutputOp>(use.getOwner())) {
+          rewriter.updateRootInPlace(use.getOwner(), [&]() { use.set(sig); });
+        }
       }
+
+      // Probe the value of the signal such that we end up having a replacement
+      // for the InstanceOp results later on.
+      auto prb = rewriter.create<PrbOp>(loc, resultType, sig);
+      prb->setAttr("magic2", UnitAttr::get(rewriter.getContext()));
```

Comment:
Is this just for debug?

Reply:
Yeah good catch. Will drop that, thanks.